Mathlib/Data/Matrix/Basic.lean

```diff
@@ -0,0 +1,2829 @@
+    (diagonal d).submatrix e e = diagonal (d ∘ e) :=
+  submatrix_diagonal d e e.injective
+#align matrix.submatrix_diagonal_equiv Matrix.submatrix_diagonal_equiv
+
+@[simp]
+theorem submatrix_one_embedding [Zero α] [One α] [DecidableEq m] [DecidableEq l] (e : l ↪ m) :
+    (1 : Matrix m m α).submatrix e e = 1 :=
+  submatrix_one e e.injective
+#align matrix.submatrix_one_embedding Matrix.submatrix_one_embedding
+
+@[simp]
+theorem submatrix_one_equiv [Zero α] [One α] [DecidableEq m] [DecidableEq l] (e : l ≃ m) :
+    (1 : Matrix m m α).submatrix e e = 1 :=
+  submatrix_one e e.injective
+#align matrix.submatrix_one_equiv Matrix.submatrix_one_equiv
+
+@[simp]
+theorem submatrix_mul_equiv [Fintype n] [Fintype o] [AddCommMonoid α] [Mul α] {p q : Type _}
+    (M : Matrix m n α) (N : Matrix n p α) (e₁ : l → m) (e₂ : o ≃ n) (e₃ : q → p) :
+    M.submatrix e₁ e₂ ⬝ N.submatrix e₂ e₃ = (M ⬝ N).submatrix e₁ e₃ :=
+  (submatrix_mul M N e₁ e₂ e₃ e₂.bijective).symm
+#align matrix.submatrix_mul_equiv Matrix.submatrix_mul_equiv
+
+theorem mul_submatrix_one [Fintype n] [Fintype o] [NonAssocSemiring α] [DecidableEq o] (e₁ : n ≃ o)
+    (e₂ : l → o) (M : Matrix m n α) :
+    M ⬝ (1 : Matrix o o α).submatrix e₁ e₂ = submatrix M id (e₁.symm ∘ e₂) := by
+  let A := M.submatrix id e₁.symm
+  have : M = A.submatrix id e₁ := by
+    simp only [submatrix_submatrix, Function.comp.right_id, submatrix_id_id, Equiv.symm_comp_self]
+  rw [this, submatrix_mul_equiv]
+  simp only [Matrix.mul_one, submatrix_submatrix, Function.comp.right_id, submatrix_id_id,
+    Equiv.symm_comp_self]
+#align matrix.mul_submatrix_one Matrix.mul_submatrix_one
+
+theorem one_submatrix_mul [Fintype m] [Fintype o] [NonAssocSemiring α] [DecidableEq o] (e₁ : l → o)
+    (e₂ : m ≃ o) (M : Matrix m n α) :
+    ((1 : Matrix o o α).submatrix e₁ e₂).mul M = submatrix M (e₂.symm ∘ e₁) id := by
+  let A := M.submatrix e₂.symm id
+  have : M = A.submatrix e₂ id := by
+    simp only [submatrix_submatrix, Function.comp.right_id, submatrix_id_id, Equiv.symm_comp_self]
+  rw [this, submatrix_mul_equiv]
+  simp only [Matrix.one_mul, submatrix_submatrix, Function.comp.right_id, submatrix_id_id,
+    Equiv.symm_comp_self]
+#align matrix.one_submatrix_mul Matrix.one_submatrix_mul
+
+/-- The natural map that reindexes a matrix's rows and columns with equivalent types is an
+equivalence. -/
+def reindex (eₘ : m ≃ l) (eₙ : n ≃ o) : Matrix m n α ≃ Matrix l o α where
+  toFun M := M.submatrix eₘ.symm eₙ.symm
+  invFun M := M.submatrix eₘ eₙ
+  left_inv M := by simp
+  right_inv M := by simp
+#align matrix.reindex Matrix.reindex
+
+@[simp]
+theorem reindex_apply (eₘ : m ≃ l) (eₙ : n ≃ o) (M : Matrix m n α) :
+    reindex eₘ eₙ M = M.submatrix eₘ.symm eₙ.symm :=
+  rfl
+#align matrix.reindex_apply Matrix.reindex_apply
+
+-- @[simp] -- Porting note: simp can prove this
+theorem reindex_refl_refl (A : Matrix m n α) : reindex (Equiv.refl _) (Equiv.refl _) A = A :=
```

Comment:
Whoops, this one was probably meant to say something else. That's out of scope here though.